src/adrs/0129349218/README.md

```diff
@@ -0,0 +1,574 @@
+  domain of errors.
+- `id`: The ID of an instance of the error. It could be a UUID, Nanoid, or any
+  other unique identifier that could be generated without coordination.
+
+When dealing Error Aggregation and Analysis the `domain` and `reason` fields
+will be used to identify the context in which the error happened and the `id`
+field will be used to identify the instance of the error.
+
+For example, you may decide that a `not_found` `reason` is will be used across
+all the domains. Such `reason` by itself is too generic, and you couldn't tell
+where the error happened, but when you combine it with the `domain` you can
+identify the context in which the error happened.
+
+For example, having the following error:
+
+```js
+const error = {
+  domain: "com.app.bank_transfer",
+  reason: "not_found",
+  data: {transfer_id: "709b4d54-04ee-4e82-89a3-4bdf07462809"}
+}
+```
+
+You can now identify that the error happened in the `com.app.bank_transfer`
+service (or domain), and the resource that was not found was the transfer with
+the id `709b4d54-04ee-4e82-89a3-4bdf07462809`.
+
+If you wish to add have some infrastructure tooling that map `not_found` to a
+404 HTTP status, you can do so without having to parse the error type, allowing
+you to have more flexibility in the future.
+
+Standardizing the `domain` and `reason` fields will allow you to have a common
+way to document errors across multiple systems and also have a common way to
+generate documentation for them. Such standardization is out of scope for this
+specification.
+
+#### Documentation
+
+Errors are part of the API surface, therefore, they should be documented.
+Having a common structure for errors will allow you to have a common way to
+document them. This will allow you to have a common way to document errors
+across multiple systems and also have a common way to generate documentation for
+them.
+
+Adding a `message` to make it more human-readable for programmers to understand
+the error and `info.metadata` to provide context to the error message.
+
+The error should have a `help` field that points to more detailed documentation
+about the error when possible.
+
+#### Localization
+
+> Note: details about translation services are out of scope for this
+> specification.
+
+Errors that are meant to be displayed to the end-user should be localized.
+
+Using the `domain`, `reason` and `metadata` fields should allow to translate the
+error message to the end-user language. Using `domain` and `reason` to identify
+the error and `metadata` to provide the context to the error message.
+
+For example, imagine the following translation data:
+
+```js
+const translations = {
+  "com.myapp.bank_transfer": {
+    "transfer_limit": "You have reached the transfer limit of {{transfer_limit}}."
+  }
+}
+```
+
+When you receive the following error, you could translate it to the end-user
+language:
+
+```js
+const error = {
+  info: {
+    domain: 'com.myapp.bank_transfer',
```

Comment:
One thing that might sound overkill ( but we're not considering ourselves with network performance ) here is this. Because a domain/reason combo might not always be unique i always give each error a unique id in the codepath that makes me find instantly where the error originated from.

I understand why some people don't want to do that and consider it dirty but it has never failed to help me instantly identify the codepath with a single search. So in this case this would be something like err_id: 59314 which would also be hardcoded on the b.e

Reply:
Mentioned this in Discord but I believe an optional error code would be a good addition. There's many existing systems/organizations that utilize numeric error codes and you can find similar recommendations in various resources on the topic.

Reply:
As we discussed,

Yhe combination of domain and reason is that. Having the information to be two dimensions allow us to define error codes at the Product and Platform level a bit easier
while the metadata gives the 3rd dimension to make sense.

Example:

```
domain: googleapi
reason: not_found
metadata: { resource_id, resource_type }
```